File: 257-binary-tree-paths/binary-tree-paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution 
{
    public:
        void findAllPaths(TreeNode* root,vector<string>&ans,string curr)
        {
            if(root==NULL)
            {
                return;
            }

            string newCurr;
            if(curr=="")
            {
                newCurr=to_string(root->val);
            }
            else
            {
                newCurr=curr+"->"+to_string(root->val);
            }

            if(root->left==NULL && root->right==NULL)
            {
                ans.push_back(newCurr);
                return;
            }
            findAllPaths(root->left,ans,newCurr);
            findAllPaths(root->right,ans,newCurr); 
        }
        vector<string> binaryTreePaths(TreeNode* root) 
        {
            vector<string>ans;
            string curr;

            findAllPaths(root,ans,curr);
            return ans;
        }
};

File: 257-binary-tree-paths/binary-tree-paths.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void findPaths(TreeNode root,StringBuilder str,List<String>ans)
    {
        if(root==null)
        {
            return;
        }
        int originalLength=str.length();
        if(originalLength>0)
        {
            str.append("->");
        }
        str.append(root.val);

        if(root.left==null && root.right==null)
        {
            ans.add(str.toString());
        }
        else
        {
            findPaths(root.left,str,ans);
            findPaths(root.right,str,ans);
        }
        str.setLength(originalLength);
    }
    public List<String> binaryTreePaths(TreeNode root) 
    {
        List<String>ans=new ArrayList();
        StringBuilder str=new StringBuilder();

        findPaths(root,str,ans);
        return ans;
    }
}